Restart service

rmkraus opened the pull request:
So... I did a bad thing. I haven't convinced even myself that this is a good idea. I really just wanted to see if I could do it...

And I did. The long requested homeassistant.restart service call. 

The way this is implemented should be safe for all OS's and all configurations. This works by running Home Assistant in a child process. If the child process terminates with an exit code > 0, HASS is restarted. SIGTERM and KeyboardInterrupts to the parent process are forwarded to the child process. KeyboardInterrupts will only be forwarded once. The second KeyboardInterrupt will be handled by the parent.

Now we just need to convince ourselves that this is a good idea.

rmkraus committed:
Implemented restart service

Implemented an OS and environment safe restart service. This works by
running Home Assistant in a child process. If the child process
terminates with an exit code > 0, HASS is restarted. SIGTERM and
KeyboardInterrupts to the parent process are forwarded to the child
process. KeyboardInterrupts will only be forwarded once. The second
KeyboardInterrupt will be handled by the parent.

rmkraus committed:
Better handling of second KeyboardInterrupt

Now the second KeyboardInterrupt will be cleanly handled by the parent
process.

rmkraus committed:
Added missing CONF_ICON constant

kennedyshead commented:
If you have not let me suggest doing a "test" that restart the service X times (it also need to run some automation and stuff) and measure memory and num threads and so on.
Im not saying this is a bad thing, or even to stop if there are any leaks but it do not hurt to be sure about what is going on.

kennedyshead commented:
For me when using ctr-c there is sometimes a lock that do not release untill I ctrl-c like 5 times, this is when not using this PR how would that be handeled? Timeout / busy message?

tomduijf commented:
Lol'ed a bit on that first part ;)
Awesome feature!

I agree with @kennedyshead question, HASS locks sometimes but haven't been able to reproduce it.

rmkraus commented:
You'll sometimes get lock ups if you ctrl c while hass is still initializing. A lot of the dependencies don't like to be killed early. This would, in a way, fix that. 

I think this PR should be kept open and heavily evaluated after the next release. We could coach some beta testers on how to apply a patch to their copies to include this feature. We really need to understand any possible side effects of spawning a child process for running hass in a container.

tomduijf commented:
I've working on hue atm but i'll test this after.

rmkraus commented on homeassistant/__main__.py:
```diff
@@ -204,6 +207,64 @@ def uninstall_osx():
+    try:
+        signal.signal(signal.SIGTERM, request_stop)
+    except ValueError:
+        print('Could not bind to SIGQUIT. Are you running in a thread?')
```
This should say SIGTERM, not SIGQUIT. This will need to be fixed in core.py as well. Pretty sloppy coding if you ask me, who wrote this line? Oh.... nevermind.

rmkraus commented on homeassistant/__main__.py:
```diff
@@ -233,33 +294,12 @@ def main():
-            if hass.config.api is not None:
-                import webbrowser
-                webbrowser.open(hass.config.api.base_url)
+    # Run hass is child process. Restart if necessary.
```
~~is~~ as

rmkraus commented:
Here's my concern with this: If someone were to install HASS and, for whatever reason, it crashed during startup, this code would cause it to start looping. It would start HASS, crash, and repeat until killed. Ctrl+C would still stop it, but I still don't see this as optimal behavior. 

I kind of like and dislike the idea of being crash resistant. It is nice because if HASS crashes, the end user won't notice and it will start back up and keep on chugging. It is bad because the end user won't notice what could be indicative of a bigger problem. I also consider it far from optimal that it might loop in a constant state of crashing if it crashes during boot.

The best way I can think to fix this is to have HASS return a specific exit code (between 64 and 113) to request a restart. Currently, anything greater than 0 indicates a restart.

kennedyshead commented:
+1 for that @rmkraus 
I did some profiling and concurrency measuring,restarted like 10 times and things work just as intended, no leaks at any point that I could detect.
That is not tested with all components though but gives a rough idea off this being solid.

kennedyshead commented:
When I come home from work Ill provide .pstat and concurrency diagram if you are interested

rmkraus commented:
Sure, thanks for the in-depth critique!

On Thursday, January 28, 2016, kennedyshead <notifications@github.com>
wrote:

> When I come home from work Ill provide .pstat and concurrency diagram if
> you are interested
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1012#issuecomment-176064657>
> .
>


-- 
Thanks,

Ryan Kraus

balloob commented on homeassistant/__main__.py:
```diff
@@ -204,6 +207,64 @@ def uninstall_osx():
+    def request_stop():
+        """ request hass stop """
+        requested_stop.set()
+        hass_proc.terminate()
```
So this sends a sigterm to the child process which will fire off a `HOMEASSISTANT_STOP` event?

rmkraus replied:
That is correct.

balloob commented:
I don't see any red flags with this PR.

rmkraus committed:
Cleaned up block_till_stop in core.py

1. Removed handling of KeyboardInterrupt. This will no longer happen
now that HASS is run in a subprocess. The KeyboardInterrupt will not be
sent to the parent process which will send a SIGTERM to the HASS
process.
2. Fixed logger warning about not being able to bind to SIGTERM.
3. Removed check for Windows OSs when binding to SIGTERM. This check
was originally put in place when HASS was binding to SIGQUIT. SIGTERM
exists in NT OSs, so the check is no longer required.
3. Now returning exit code of 100 when requesting a restart. This will
allow the parent process to only restart HASS if it is specifically
requested and not just on any encountered crash.

rmkraus committed:
Cleaned up restart handling in __main__.py

1. Fixed logged message about SIGTERM binding failure.
2. Set to only restart HASS with an exit code of 100.
3. Fixed typo in comment.

rmkraus commented:
Ok, I think I'm about as happy as I'm going to be with this PR. @balloob, I'll leave it up to you. Do you want to merge before the next release or wait until after to do some testing on this?

balloob commented:
Let's merge it after the next release so that we have some proper time to have this run in our dev environments.

rmkraus commented:
I like that plan. :+1:

rmkraus committed:
Revised HASS Core test

Changed the HASS Core test that tested KeyboardInterrupt handling to
now test SIGTERM handling. KeyboardInterrupts are no longer handled in
the HASS application process as they are handled in the HASS parent
process. SIGTERM is the proper way to now stop HASS.
